diff --git a/depends/CMakeLists.txt b/depends/CMakeLists.txt
index fa8ccef..98a8078 100755
--- a/depends/CMakeLists.txt
+++ b/depends/CMakeLists.txt
@@ -3,9 +3,19 @@ if(${CURVE} STREQUAL "BN128")
   include_directories(xbyak)
   add_library(
     zm
-    STATIC
+    SHARED
 
     ate-pairing/src/zm.cpp
     ate-pairing/src/zm2.cpp
   )
+
+  install(
+    DIRECTORY "" DESTINATION "include/depends"
+    FILES_MATCHING
+    PATTERN "*.h"
+  )
+
+  install(
+    TARGETS zm DESTINATION lib
+  )
 endif()
diff --git a/libff/CMakeLists.txt b/libff/CMakeLists.txt
index bcc0bb0..374f335 100755
--- a/libff/CMakeLists.txt
+++ b/libff/CMakeLists.txt
@@ -22,7 +22,7 @@ endif()
 
 add_library(
   ff
-  STATIC
+  SHARED
 
   algebra/curves/alt_bn128/alt_bn128_g1.cpp
   algebra/curves/alt_bn128/alt_bn128_g2.cpp
